python/thunder/rdds/data.py

```diff
@@ -313,12 +313,29 @@ def astype(self, dtype, casting='safe'):
         """
         if dtype is None or dtype == '':
             return self
+        from numpy import can_cast, ndarray
+        from numpy import dtype as dtypeFunc
         if dtype == 'smallfloat':
             # get the smallest floating point type that can be safely cast to from our current type
             from thunder.utils.common import smallestFloatType
             dtype = smallestFloatType(self.dtype)
 
-        nextRdd = self.rdd.mapValues(lambda v: v.astype(dtype, casting=casting, copy=False))
+        def cast(v, dtype_, casting_):
+            if isinstance(v, ndarray):
+                return v.astype(dtype_, casting=casting_, copy=False)
+            elif can_cast(v, dtype_, casting=casting_):
+                if hasattr(v, 'astype'):
+                    # numpy scalars have an 'astype' method, but don't have the "casting" kwarg:
+                    return v.astype(dtype_)
+                else:
+                    # turn ourself into a numpy scalar of the appropriate type
+                    # despite the name, 'asarray' will return a scalar if a scalar is passed
+                    return asarray(v, dtype=dtype_)
```

Comment:
Discussed offline with @industrial-sloth , we're going to simplify this by dropping the additional `hasattr` type, and just call `return asarray([v], dtype=dtype_)[0]`, which will convert to the appropriate type and return a valid singleton `numpy` object.